src/controller.cpp

```diff
@@ -156,27 +181,34 @@ void Controller::setPlan(geometry_msgs::Transform current_tf, geometry_msgs::Twi
   /* Minimal sanity check */
   global_plan_tf_.clear();
   global_plan_tf_.push_back(to_transform(global_plan[0].pose));
+
   // For now do not allow repeated points or in-place rotation
   // To allow that the way the progress is checked and the interpolation is done needs to be changed
   // Also check if points suddenly go in the opposite direction, this could lead to deadlocks
-  for (std::vector<geometry_msgs::PoseStamped>::size_type pose_idx = 1; pose_idx < global_plan.size() - 1; ++pose_idx)
-  {
-    const auto prev_pose = global_plan[pose_idx - 1].pose;
-    const auto pose = global_plan[pose_idx].pose;
-    const auto next_pose = global_plan[pose_idx + 1].pose;
-    if (is_pose_angle_obtuse(prev_pose, pose, next_pose))
-    {
-      global_plan_tf_.push_back(to_transform(pose));
-    }
-    else
-    {
-      ROS_WARN("Pose %zu of path is not used since it is not in the expected direction of the path!", pose_idx);
-    }
+  if (global_plan.size() > 2) {
+    const auto it_begin = global_plan.cbegin();
+    const auto it_end = global_plan.cend() - 2;
+    const auto if_predicate = [](const auto & prev, const auto & cur, const auto & next) {
+      return is_pose_angle_obtuse(prev.pose, cur.pose, next.pose);
+    };
+    const auto transform_predicate = [](
+                                       const auto & /* prev */, const auto & cur,
+                                       const auto & /* next */) { return to_transform(cur.pose); };
+
+    transform_if(
+      it_begin, it_end, it_begin + 1, it_begin + 2, std::back_inserter(global_plan_tf_),
+      if_predicate, transform_predicate);
```

Comment:
I can understand the motivation for this change, however I would do it slightly different.

In the future we want all internal datatypes to be `tf2` so that we can immediately can perform calculations with them. This will make the `to_transform` call not needed.

As an intermediate step we could split this algorithm in to
```
std::transform(...)  // convert path into tf2 datatypes
std::copy_if(...is_pose_angle_obtuse(...)) // copy only obtuse angles
```
Later we could then remove the `std::transform` call.

What do you think about this method?